substratecommon.go

```diff
@@ -0,0 +1,817 @@
+
+// Transaction represents summary information about a transaction.
+type Transaction struct {
+	ID          string
+	Reason      string
+	Event       []byte
+	ChaincodeID string
+}
+
+// Block represents summary information about a block.
+type Block struct {
+	Hash         string
+	Transactions []*Transaction
+}
+
+// Substrate is the interface that we're exposing as a plugin.
+type Substrate interface {
+	NewRPC() (string, error)
+	CloseRPC(string) error
+
+	NewMockFrom(string, string, []byte) (string, error)
+	SetCreatorWithAttributesMock(string, string, map[string]string) error
+	SnapshotMock(string) ([]byte, error)
+	CloseMock(string) error
+
+	Init(string, string, *ConcreteRequestOptions) error
+	Call(string, string, *ConcreteRequestOptions) (*Response, error)
+	QueryInfo(string, *ConcreteRequestOptions) (uint64, error)
+	QueryBlock(string, uint64, *ConcreteRequestOptions) (*Block, error)
+
+	// IsTimeoutError doesn't use RPC
+	IsTimeoutError(err error) bool
+}
+
+// ArgsNewRPC encodes the arguments to NewRPC
+type ArgsNewRPC struct {
+}
+
+// RespNewRPC encodes the response from NewRPC
+type RespNewRPC struct {
+	Tag string
+	Err *Error
+}
+
+// ArgsCloseRPC encodes the arguments to CloseRPC
+type ArgsCloseRPC struct {
+	Tag string
+}
+
+// RespCloseRPC encodes the response from CloseRPC
+type RespCloseRPC struct {
+	Err *Error
+}
+
+// ArgsNewMockFrom encodes the arguments to NewMockFrom
+type ArgsNewMockFrom struct {
+	Name     string
+	Version  string
+	Snapshot []byte
+}
+
+// RespNewMockFrom encodes the response from NewMockFrom
+type RespNewMockFrom struct {
+	Tag string
+	Err *Error
+}
+
+// ArgsSetCreatorWithAttributesMock encodes the arguments to SetCreatorWithAttributesMock
+type ArgsSetCreatorWithAttributesMock struct {
+	Tag     string
+	Creator string
+	Attrs   map[string]string
+}
+
+// RespSetCreatorWithAttributesMock encodes the response from SetCreatorWithAttributesMock
+type RespSetCreatorWithAttributesMock struct {
+	Err *Error
+}
+
+// ArgsSnapshotMock encodes the arguments to SnapshotMock
+type ArgsSnapshotMock struct {
+	Tag string
+}
+
+// RespSnapshotMock encodes the response from SnapshotMock
+type RespSnapshotMock struct {
+	Snapshot []byte
+	Err      *Error
+}
+
+// ArgsCloseMock encodes the arguments to CloseMock
+type ArgsCloseMock struct {
+	Tag string
+}
+
+// RespCloseMock encodes the response from CloseMock
+type RespCloseMock struct {
+	Err *Error
+}
+
+// ArgsInit encodes the arguments to Init
+type ArgsInit struct {
+	Tag     string
+	Phylum  string
+	Options *ConcreteRequestOptions
+}
+
+// RespInit encodes the response from Init
+type RespInit struct {
+	Err *Error
+}
+
+// ArgsCall encodes the arguments to Call
+type ArgsCall struct {
+	Tag     string
+	Command string
+	Options *ConcreteRequestOptions
+}
+
+// RespCall encodes the response from Call
+type RespCall struct {
+	Response *Response
+	Err      *Error
+}
+
+// ArgsQueryInfo encodes the arguments to QueryInfo
+type ArgsQueryInfo struct {
```

Comment:
Do all of these structs need to be exposed publically or are they just used internally?

Reply:
Yes, I think RPC requires the fields to be public: `- the method has two arguments, both exported (or builtin) types.`